core/embed/rust/src/ui/layout_eckhart/firmware/device_menu_screen.rs

```diff
@@ -0,0 +1,542 @@
+    action: Option<Action>,
+}
+
+impl MenuItem {
+    pub fn new(text: TString<'static>, action: Option<Action>) -> Self {
+        Self {
+            text,
+            subtext: None,
+            stylesheet: theme::menu_item_title(),
+            action,
+        }
+    }
+
+    pub fn with_subtext(mut self, subtext: Option<(TString<'static>, Option<TextStyle>)>) -> Self {
+        self.subtext = subtext;
+        self
+    }
+
+    pub fn with_stylesheet(mut self, stylesheet: ButtonStyleSheet) -> Self {
+        self.stylesheet = stylesheet;
+        self
+    }
+}
+
+struct Submenu {
+    header_text: TString<'static>,
+    show_battery: bool,
+    items: Vec<MenuItem, SHORT_MENU_ITEMS>,
+}
+
+impl Submenu {
+    pub fn new(header_text: TString<'static>, items: Vec<MenuItem, SHORT_MENU_ITEMS>) -> Self {
+        Self {
+            header_text,
+            show_battery: false,
+            items,
+        }
+    }
+
+    pub fn with_battery(mut self) -> Self {
+        self.show_battery = true;
+        self
+    }
+}
+
+// Each subscreen of the DeviceMenuScreen is one of these
+enum Subscreen {
+    // A registered submenu
+    Submenu(usize),
+
+    // A screen allowing the user to to disconnect a device
+    DeviceScreen(
+        TString<'static>, /* device name */
+        usize,            /* index in the list of devices */
+    ),
+
+    // The about screen
+    AboutScreen,
+}
+
+pub struct DeviceMenuScreen<'a> {
+    bounds: Rect,
+
+    battery_percentage: u8,
+    firmware_version: TString<'static>,
+
+    // These correspond to the currently active subscreen,
+    // which is one of the possible kinds of subscreens
+    // as defined by `enum Subscreen` (DeviceScreen is still a VerticalMenuScreen!)
+    // The active one will be Some(...) and the other one will be None.
+    // This way we only need to keep one screen at any time in memory.
+    menu_screen: GcBox<Option<VerticalMenuScreen<ShortMenuVec>>>,
+    about_screen: GcBox<Option<TextScreen<Paragraphs<[Paragraph<'a>; 2]>>>>,
+
+    // Information needed to construct any subscreen on demand
+    submenus: GcBox<Vec<Submenu, MAX_SUBMENUS>>,
+    subscreens: Vec<Subscreen, MAX_SUBSCREENS>,
+
+    // index of the current subscreen in the list of subscreens
+    active_subscreen: usize,
+
+    // stack of parents that led to the current subscreen
+    parent_subscreens: Vec<usize, MAX_DEPTH>,
+}
+
+impl<'a> DeviceMenuScreen<'a> {
+    pub fn new(
+        failed_backup: bool,
+        battery_percentage: u8,
+        firmware_version: TString<'static>,
+        device_name: TString<'static>,
+        // NB: we currently only support one device at a time.
+        // if we ever increase this size, we will need a way to return the correct
+        // device index on Disconnect back to uPy
+        // (see component_msg_obj.rs, which currently just returns "DeviceDisconnect" with no
+        // index!)
+        paired_devices: Vec<TString<'static>, 1>,
```

Comment:
Do we want to use a constant for this `Vec` size?
Or maybe define it as a `PairedDeviceVec<T>`?